Get normalized scores

rodrigodelazcano opened the pull request:
# Description

This PR adds the [`get_normalized_score()` function from D4RL](https://github.com/Farama-Foundation/D4RL/blob/71a9549f2091accff93eeff68f1f3ab2c0e0a288/d4rl/offline_env.py#L71) for policy testing and comparisson. This function is added as a utility instead of a method in every MinariDataset since we haven't decided if we want to keep this feature for all future datasets in Minari and not all offline RL tasks can by tested in this manner (due to complexity of generating an expert policy for a reference maximum score).

This function requires the dataset to have two attributes which can be set when creating the dataset, `ref_min_score` and `ref_max_score`. These attributes are optional and if not present in the dataset `get_normalized_socre()` will through an error.
To set these attributes they can be directly passed by the user when creating the dataset, or by passing a callable expert policy to obtain the episode returns from. Note that if `ref_min_score` is not passed, it will be internally computed with a random policy.

I'm using this PR to also fix an issue with storing dictionary observation spaces in the DataCollectorV0 and not resetting the environment when `terminated` or `truncated`. I added a comment in the specific line changes.

This PR is a Draft because I need to update the PointMaze dataset generation tutorial to include the reference scores. There are some issues and differences with the D4RL PointMaze reference scores that I'll list here.

## Type of change

> Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

### Screenshots

> Please attach before and after screenshots of the change if applicable.
> To upload images to a PR -- simply drag and drop or copy paste.

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have run `pytest -v` and no errors are present.
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I solved any possible warnings that `pytest -v` has generated that are related to my code to the best of my knowledge.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

<!--
You can mark something as done by putting an x character in it

For example,
- [x] I have done this task
- [ ] I have not done this task
-->

rodrigodelazcano commented on minari/data_collector/data_collector.py:
```diff
@@ -230,7 +230,12 @@ def step(
         # certain conditions.
         if self._new_episode and not self._reset_called:
-            self._buffer[-1]["observations"] = [self._previous_eps_final_obs]
+            if isinstance(self._previous_eps_final_obs, dict):
```
Fix for dictionary observation spaces

balisujohn replied:
If possible, I think it would be good to add a test for this behavior.

younik commented on minari/utils.py:
```diff
@@ -333,18 +414,76 @@ def create_dataset_from_collector_env(
+    Returns:
+        normalized_scores
+    """
+    with h5py.File(dataset.spec.data_path, "r") as f:
+        ref_min_score = f.attrs.get("ref_min_score", default=None)
+        ref_max_score = f.attrs.get("ref_max_score", default=None)
+    if ref_min_score is None or ref_max_score is None:
+        raise ValueError(
+            f"Reference score not provided for dataset {dataset.spec.dataset_id}. Can't compute the normalized score."
```
Maybe it is better to access those attributes through `MinariStorage`
Should we make the `MinariDataset._data` gettable?

rodrigodelazcano replied:
The idea here is to avoid extra optional attributes in the MinariStorage and MinariDataset. What we can do is move this function to the MinariStorage (that uses h5py`), and then call this function throught another method in MinariDataset. 
```python
def get_normalized_score(self, score):
    return self._data.get_normalized_score(score)
```

younik replied:
> The idea here is to avoid extra optional attributes in the MinariStorage and MinariDataset. What we can do is move this function to the MinariStorage (that uses h5py`), and then call this function throught another method in MinariDataset.
> 
> ```python
> def get_normalized_score(self, score):
>     return self._data.get_normalized_score(score)
> ```
> 
> Do you think this is a better option? In my opinion I prefer the current implementation to keep this feature as separated as possible from the dataset structures until we decide if we are going to maintain it at all.

I prefer the current implementation. I was also thinking on a general method `get_attr` to get some attribute from the data (and throw an error if there is no such attribute)

rodrigodelazcano replied:
I like the idea of `get_attr` can you add this in a separe PR @younik?

Howuhh commented on minari/utils.py:
```diff
@@ -197,6 +239,11 @@ def create_dataset_from_buffers(
         author (Optional[str], optional): author that generated the dataset. Defaults to None.
         author_email (Optional[str], optional): email of the author that generated the dataset. Defaults to None.
         code_permalink (Optional[str], optional): link to relevant code used to generate the dataset. Defaults to None.
+        ref_min_score( Optional[float], optional): minimum reference score from the average returns of a random policy. This value is later used to normalize a score with :meth:`minari.get_normalized_score`. If default None the value will be estimated with a default random policy.
```
should we even compute min score if max score is not provided? we won't be able to use it anyway

rodrigodelazcano replied:
This is already being done. First checking if an `expert_policy` or `ref_max_score` is passed and then computing the respective attributes. I'll update the docs, I agree it's not very clear that functionality

rodrigodelazcano commented on minari/utils.py:
```diff
@@ -333,18 +414,76 @@ def create_dataset_from_collector_env(
+            "code_permalink": str(code_permalink),
+        }
+
+        if expert_policy is not None or ref_max_score is not None:
+            env = copy.deepcopy(collector_env.env)
+            if ref_min_score is None:
+                ref_min_score = get_average_reference_score(
+                    env, RandomPolicy(env), num_episodes_average_score
+                )
+
+            if expert_policy is not None:
+                ref_max_score = get_average_reference_score(
+                    env, expert_policy, num_episodes_average_score
+                )
```
Sounds good, I'll update this section and then we can look into making other refactors

rodrigodelazcano commented:
> Are we committing to at least leaving a version of `get_normalized_scores` in the codebase? If so, I think it would be good to write some sort of test for it. If not, then I think it's fine to hold off on tests for now.

We should be keeping it in all future codebase, at least for some of the d4rl datasets. I'll add the tests :)

rodrigodelazcano commented:
I added a test for

> Are we committing to at least leaving a version of `get_normalized_scores` in the codebase? If so, I think it would be good to write some sort of test for it. If not, then I think it's fine to hold off on tests for now.

I added a test for this fix https://github.com/Farama-Foundation/Minari/pull/110#discussion_r1262846854. We can make the test for `get_normlized_score` in another PR

balisujohn commented:
LGTM :^)
